src/convnets/resnets.jl

```diff
@@ -0,0 +1,543 @@
+  - `drop_rates`: A `NamedTuple` that can may define none, some or all of the following:
+    
+      + `dropout_rate`: The rate of dropout to be used in the classifier head.
+      + `drop_path_rate`: Stochastic depth implemented using [`DropPath`](#).
+      + `drop_block_rate`: `DropBlock` regularisation implemented using [`DropBlock`](#).
+  - `classifier_args`: A `NamedTuple` that may define none, some or all of the following:
+    
+      + `pool_type`: The type of pooling to use in the classifier head. Uses
+        [`SelectAdaptivePool`](#) to select the pooling function. See its
+        documentation for more information.
+      + `use_conv`: Whether to use a 1x1 convolutional layer in the classifier head instead of a
+        `Dense` layer.
+"""
+function resnet(block_fn, layers; inchannels = 3, nclasses = 1000, output_stride = 32,
+                stem = first(resnet_stem(; inchannels)), inplanes = 64,
+                downsample_fn = downsample_conv, block_args::NamedTuple = NamedTuple(),
+                drop_rates::NamedTuple = (dropout_rate = 0.0, drop_path_rate = 0.0,
+                                          drop_block_rate = 0.0),
+                classifier_args::NamedTuple = NamedTuple())
+    # Feature Blocks
+    channels = [64, 128, 256, 512]
+    stage_blocks = _make_blocks(block_fn, channels, layers, inplanes;
+                                output_stride, downsample_fn, drop_rates, block_args)
+    # Head (Pooling and Classifier)
+    expansion = expansion_factor(block_fn)
+    num_features = 512 * expansion
+    global_pool, fc = create_classifier(num_features, nclasses; classifier_args...)
+    classifier = Chain(global_pool, Dropout(get(drop_rates, :dropout_rate, 0)), fc)
+    return Chain(Chain(stem, stage_blocks), classifier)
+end
+
+# block-layer configurations for ResNet and ResNeXt models
+const resnet_config = Dict(18 => (basicblock, [2, 2, 2, 2]),
+                           34 => (basicblock, [3, 4, 6, 3]),
+                           50 => (bottleneck, [3, 4, 6, 3]),
+                           101 => (bottleneck, [3, 4, 23, 3]),
+                           152 => (bottleneck, [3, 8, 36, 3]))
+struct ResNet
+    layers::Any
+end
+@functor ResNet
+
+(m::ResNet)(x) = m.layers(x)
+
+"""
+    ResNet(depth::Integer; pretrain = false, inchannels = 3, nclasses = 1000)
+
+Creates a ResNet model with the specified depth.
+((reference)[https://arxiv.org/abs/1512.03385])
+
+# Arguments
+
+  - `depth`: one of `[18, 34, 50, 101, 152]`. The depth of the ResNet model.
+  - `pretrain`: set to `true` to load the model with pre-trained weights for ImageNet
+  - `inchannels`: The number of input channels.
+  - `nclasses`: the number of output classes
+
+!!! warning
+    
+    `ResNet` does not currently support pretrained weights.
+
+Advanced users who want more configuration options will be better served by using [`resnet`](#).
+"""
+function ResNet(depth::Integer; pretrain = false, inchannels = 3, nclasses = 1000)
```

Comment:
We seem to have lost the ability to specify the shortcut type like we used to with the older interface.

Reply:
I looked for this in other common implementations and couldn't find it. In fact, I was rather confused with the usage of a `MaxPool` in there to allow `skip_identity` even when it wasn't an identity projection. I could maybe try and add it back, but I think the current interface allows enough flexibility with the downsampling function (`downsample_conv` and `downsample_pool`) that this should not be missed

Reply:
The max pooling was there for the case where the feature map size changed, though the logic of `outplanes > inplanes` was a brittle way for detecting that (it was closely linked to how `resnet` worked).

What's lost here is the ability to easily create a parameter-free projection which is nice for small ResNets that are being mapped to constrained hardware. I think we want one more downsampling default called `downsample_identity` that behaves like `skip_identity`.

Another piece that's missing is the ability to customize the downsampling function beyond the first repetition of a block. This eliminates the possibility of using Shortcut C. I am not sure of the use case for this one, and it isn't implemented in any other framework. The only issue is that it is a feature loss. This could be remedied by allowing `downsample_fn` to be a 2-tuple like in the old API.

Reply:
Ah, fair enough. I'll work on making this happen

Reply:
Okay so the main issue with making this happen is the `Vector` of choices for the stages. Earlier there were just three choices and so they were hardcoded, but now we have around 6-7 choices and having an alphabet for each one doesn't seem the best way to go. What do you suggest I do for that case?

Reply:
I am not sure I follow. The number of stages and the types of shortcuts are orthogonal? There should still only be three shortcut types.

Reply:
There's an additional `downsample_pool` option now as well 😅

Reply:
Ah, I see. The alphabet naming corresponds directly to the paper. The alphabet based constructor is only for maintaining an existing feature, so I don't think we need to extend it to the new features added. Those can be specified as `(downsample_pool, downsample_identity)` by the user manually.

Reply:
Yep but that's the problem - for the `resnet` constructor, we were passing in a `Vector` of symbols. For the new API, does it become a `Vector` of `Tuples` with the functions (in case it's something custom?)

Reply:
The main builder would accept a vector of tuples of functions (as did the old code). We would add a dispatch for `ResNet` that builds the vector of tuples [like we did before](https://github.com/FluxML/Metalhead.jl/blob/b37bee7eb0a26b5fd078d59ae568ce1dda6678a6/src/convnets/resnet.jl#L146-L155).

Reply:
That makes sense. I'm not sure it's a great idea to expose this at the `ResNet` level, though. How about `resnet` has the dispatch for both the `Vector` of `Symbol`s and the `Vector` of `Tuple`s with the functions? The user can decide which one to use. It does make `resnet` a little cluttered but I was trying to make `ResNet` as clean as possible so that there's a really quick start possible

Reply:
The main reason for supporting the symbols was that we have an exported dispatch for `ResNet` that allows them. Otherwise, we need to release a patch with a deprecation warning first. At the `resnet` level, I think just the vector of tuples of functions is fine to reduce clutter as you mentioned.

Reply:
Though I don't have a preference where we intercept the symbols as long as the `ResNet` constructor with symbols is handled somehow (or we patch release).

Reply:
I think a patch might have to be considered in the v0.7 series. There's too many changes here to be considered even mildly backwards-compatible (`resnet` alone, although not exported, has been dismantled completely and it was not private so I'm sure some users _have_ used it somewhere)

Reply:
`resnet` was not exported so I don't think we need to deprecate it in any way so long as we release the new code in a breaking release. The main interfaces `ResNet`, etc. seem to be done in a non-breaking manner with the exception of the shortcut constructor.

Reply:
Maybe @ToucheSir can weight in and we can just pick a path forward.

Reply:
Sorry, what would you like me to weigh in on? It does look like we're on the hook for `resnet` because it's in the docs here: https://fluxml.ai/Metalhead.jl/0.7.3/docstrings/Metalhead.resnet.html.

Reply:
That will mean a deprecation notice in a patch release, correct? Might be good to keep that waiting because I anticipate the ViTs breaking quite a bit too (not to mention trying to unify the EfficientNet interface for v2 to be added)